feat: add sync setting parameter to index

Vits-99 opened the pull request:
### **User description**
New implementation at startup to synchronize the local index with the remote index based on the specified mode.
Modes:
- `error`: Raise an error if local and remote are not synchronized.
- `remote`: Take remote as the source of truth and update local to align.
- `local`: Take local as the source of truth and update remote to align.
- `merge-force-remote`: Merge both local and remote taking only remote routes utterances when a route with same route name is present both locally and remotely.
- `merge-force-local`: Merge both local and remote taking only local routes utterances when a route with same route name is present both locally and remotely.
- `merge`: Merge both local and remote, merging also local and remote utterances when a route with same route name is present both locally and remotely.

Can be tested with basic code like the following:

```
from semantic_router import Route
from semantic_router.encoders import OpenAIEncoder
from semantic_router.index.pinecone import PineconeIndex
from semantic_router.layer import RouteLayer


politics = Route(
    name="politics",
    utterances=[
        "isn't politics the best thing ever",
        "why don't you tell me about your political opinions",
        "don't you just love the president",
    ],
)

chitchat = Route(
    name="chitchat",
    utterances=[
        "how's the weather today?",
        "how are things going?",
    ],
)

routes = [politics, chitchat]

encoder = OpenAIEncoder(openai_api_key=openai_api_key)

pc_index = PineconeIndex(
    api_key=pinecone_api_key,
    region="us-east-1",
    index_name="alai-clearwater-routes-input-intent",
    sync="local",
)

rl = RouteLayer(encoder=encoder, routes=routes, index=pc_index)
```


___

### **PR Type**
Enhancement


___

### **Description**
- Added `sync` parameter to `BaseIndex`, `LocalIndex`, and `PineconeIndex` classes.
- Introduced `_sync_index` method with various synchronization modes in `BaseIndex` and implemented it in `PineconeIndex`.
- Modified `add` method in `PineconeIndex` to handle synchronization.
- Added `_get_routes_with_ids` method in `PineconeIndex`.
- Enabled synchronization in `RouteLayer`'s `_add_routes` method.


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>base.py</strong><dd><code>Add sync parameter and synchronization logic to BaseIndex</code></dd></summary>
<hr>

semantic_router/index/base.py

<li>Added <code>sync</code> parameter to <code>BaseIndex</code> class.<br> <li> Introduced <code>_sync_index</code> method with various synchronization modes.<br>


</details>


  </td>
  <td><a href="https://github.com/aurelio-labs/semantic-router/pull/344/files#diff-df154312e842c3302d871397cc51942d0a6e3be19183c0b362499fb8e2fcd127">+21/-1</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>local.py</strong><dd><code>Add sync parameter to LocalIndex add method</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

semantic_router/index/local.py

- Added `sync` parameter to `add` method.



</details>


  </td>
  <td><a href="https://github.com/aurelio-labs/semantic-router/pull/344/files#diff-3f7c4d16e4e8638997135efd56b4c4423ca8bd5f111901351d37177ca40dfe98">+5/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pinecone.py</strong><dd><code>Add sync parameter and synchronization logic to PineconeIndex</code></dd></summary>
<hr>

semantic_router/index/pinecone.py

<li>Added <code>sync</code> parameter to <code>PineconeIndex</code> class.<br> <li> Implemented <code>_sync_index</code> method for PineconeIndex.<br> <li> Modified <code>add</code> method to handle synchronization.<br> <li> Added <code>_get_routes_with_ids</code> method.<br>


</details>


  </td>
  <td><a href="https://github.com/aurelio-labs/semantic-router/pull/344/files#diff-c57f54339773ad317ee6eafd071395202933f9eeaf59a10c28fc22d06bf564a2">+125/-1</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>layer.py</strong><dd><code>Enable synchronization in RouteLayer's _add_routes method</code></dd></summary>
<hr>

semantic_router/layer.py

- Updated `_add_routes` method to enable synchronization.



</details>


  </td>
  <td><a href="https://github.com/aurelio-labs/semantic-router/pull/344/files#diff-e495ddf91e18dd7477eb129fd2c0ab7dfbaf2440534c28b6156a76acdaf51433">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

jamescalam commented:
@Vits-99 I made some small changes to keep the qdrant index supported — wdyt to removing the `sync` parameter from the `add` method in the index objects

Vits-99 commented:
@jamescalam fixed the _get_all method ordered return bug for ids and metadata and added an _add_and_sync method to replaced the usage of the add method. I also adjusted the Qdrant index accordingly. Waiting for your feedback on this. If everything is fine I think we can merge it to main.

jamescalam reviewed (approved):
this is great @Vits-99, thanks!
